File: Twitter.cpp
//time: O(nlogk)
//space: O(k)

class Tweet {
public:
    int id, timeStamp;

    Tweet(int id, int timeStamp) {
        this->id = id;
        this->timeStamp = timeStamp;
    }
};

class Twitter {
private:
    unordered_map<int, unordered_set<int>> followedMap;
    unordered_map<int, vector<Tweet>> tweetsMap;
    int time;
    int MAX_TWEETS = 10;

public:
    Twitter()  {
        this->time = 1;
    }

    void postTweet(int userId, int tweetId) {
        if (!followedMap.count(userId)) {
            follow(userId, userId);
        }

        if (!tweetsMap.count(userId)) {
            tweetsMap[userId] = std::vector<Tweet>();
        }
        tweetsMap[userId].emplace_back(Tweet(tweetId, time++));
    }

    vector<int> getNewsFeed(int userId) {
        auto compare = [](const Tweet& a, const Tweet& b) {
            return a.timeStamp > b.timeStamp;
        };
        priority_queue<Tweet, vector<Tweet>, decltype(compare)> minHeap(compare);
        vector<int> result;

        if (!followedMap.count(userId))
            return result;

        const unordered_set<int>& users = followedMap[userId];
        vector<Tweet> fTweets;
        for (int user : users) {
            if (!tweetsMap.count(user))
                continue;
            const vector<Tweet>& tweets = tweetsMap[user];
            for (const Tweet& tweet : tweets) {
                fTweets.push_back(tweet);
            }
        }

        for (const Tweet& tweet : fTweets) {
            minHeap.push(tweet);
            if (minHeap.size() > MAX_TWEETS) {
                minHeap.pop();
            }
        }

        while (!minHeap.empty()) {
            result.insert(result.begin(), minHeap.top().id);
            minHeap.pop();
        }

        return result;
    }

    void follow(int followerId, int followeeId) {
        if (!followedMap.count(followerId)) {
            followedMap[followerId] = unordered_set<int>();
            followedMap[followerId].insert(followerId);
        }
        followedMap[followerId].insert(followeeId);
    }

    void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId || !followedMap.count(followerId))
            return;

        followedMap[followerId].erase(followeeId);
    }
};

File: Twitter.cs
using System.Collections.Generic;
using System.Collections;
namespace Design
{
    public class Twitter
    {
        /*
         * T.C: O(m*nLog10) where m is number of users , n is number of tweets and log 10 is becuase used the Min heap of 10 size
         * 
         */
        public class Tweet
        {
            public int Id;
            public int time;
            public Tweet(int id, int time)
            {
                this.Id = id;
                this.time = time;
            }
        }

        Dictionary<int, List<Tweet>> tweets;
        Dictionary<int, HashSet<int>> map;
        int time;

        PriorityQueue<Tweet, int> pq;

        public Twitter()
        {
            tweets = new Dictionary<int, List<Tweet>>();
            map = new Dictionary<int, HashSet<int>>();
            pq = new PriorityQueue<Tweet, int>();
        }

        public void PostTweet(int userId, int tweetId)
        {
            Tweet newtweet = new Tweet(tweetId, time++);
            if (!tweets.ContainsKey(userId))
            {
                tweets.Add(userId, new List<Tweet>());
            }
            tweets[userId].Add(newtweet);
        }

        public IList<int> GetNewsFeed(int userId)
        {
            Follow(userId, userId);
            HashSet<int> users = map[userId];
            foreach (int user in users)
            {
                if (tweets.ContainsKey(user))
                {
                    List<Tweet> allTweets = tweets[user];

                    foreach (Tweet twt in allTweets)
                    {
                        pq.Enqueue(twt, twt.time);

                        if (pq.Count > 10)
                        {
                            pq.Dequeue();
                        }
                    }
                }
            }

            IList<int> result = new List<int>();
            while (pq.Count != 0)
            {
                Tweet feed = pq.Dequeue();
                result.Insert(0, feed.Id);
            }

            return result;
        }

        public void Follow(int followerId, int followeeId)
        {
            if (!map.ContainsKey(followerId))
            {
                map.Add(followerId, new HashSet<int>());
            }

            map[followerId].Add(followeeId);
        }

        public void Unfollow(int followerId, int followeeId)
        {
            if (map.ContainsKey(followerId))
            {
                map[followerId].Remove(followeeId);
            }
        }
    }
}
